### DISCUSSION TITLE
Monte carlo simulation for a 2D problem

### DISCUSSION THREAD

**Discussion by vermaprk** [idaholab/moose#23716] [Q&A Modules: General] [answered]
I am trying to solve a simple monte carlo problem much similar to https://mooseframework.inl.gov/modules/stochastic_tools/examples/monte_carlo.html#4d89b646-f161-4c6e-b424-bfd47df0a7dd

However, my model is 2 dimensional, and I want to generate random boundary conditions at x=0 for both along individual y-nodes as well as in time. 

kindly suggest method to solve this problem. 
Thanks

**Comment by GiudGiud** [comment, marked as answer]
@grmnptr on two-dimensional stochastic analysis

**Reply by jiangwen84** [reply]
There are multiple ways to vary parameters, and for your case, [MultiAppSamplerControl](https://mooseframework.inl.gov/source/controls/MultiAppSamplerControl.html) would be a good choice.

For BCs, you can consider to define a function and use it in ```FunctionDirichletBC```. 

In your sub-app, you can define something like:

```
node1 = xxx
node2 = xxx
node3 = xxx
.....

[Function]
    [function]
       type = PiecewiseLinear
       x = 'node1_x node2_x ....'
       y = '${node1} ${node2} ${node3} ...'
    []
[]
```

**Reply by vermaprk** [reply]
Thanks for the quick reply.
I tried the suggested method. But I am not getting random boundary conditions. Here is my input files:

```
[StochasticTools]
  auto_create_executioner = false
[]

[Distributions]
  [uniform]
    type = Uniform
    lower_bound = 5
    upper_bound = 10
  []
[]

[Samplers]
  [sample]
    type = MonteCarlo
    num_rows = 1
    distributions = 'uniform'
    execute_on = 'PRE_MULTIAPP_SETUP'
    
  []
[]

[MultiApps]
  [sub2]
    type = SamplerTransientMultiApp
    sampler = sample
    input_files = 'sub2.i'
    execute_on = 'INITIAL TIMESTEP_END'
  []
[]

[Controls]
  [cmdline]
    type = MultiAppSamplerControl
    multi_app = sub2
    sampler = sample
    param_names = 'BCs/left/function'
  []
[]

[Executioner]
  type = Transient
  num_steps = 5
  dt = 0.01
[]
```

sub2.i file
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 10
  ny = 10
[]

[Variables]
  [u]
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = u
  []
  [time]
    type = TimeDerivative
    variable = u
  []
[]

[BCs]
  [left]
    type = FunctionDirichletBC
    variable = u
    boundary = left
    function = conductivity
  []
  [right]
    type = DirichletBC
    variable = u
    boundary = right
    value = 1
  []
  [top]
    type = NeumannBC
    variable = u
    boundary = top
    value = 0
  []
  [bottom]
    type = NeumannBC
    variable = u
    boundary = bottom
    value = 0
  []
[]

[Functions]
  [conductivity]
    type = PiecewiseLinear
    axis = y
    x = '0 0.5 1'
    y = '5.0 5.0 5.0'
  []
[]


[Executioner]
  type = Transient
  num_steps = 5
  dt = 0.01
  solve_type = PJFNK
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Controls]
  [stochastic]
    type = SamplerReceiver
  []
[]

[Postprocessors]
  [left_bc]
    type = PointValue
    point = '0 0 0'
    variable = u
  []
  [left_bc5]
    type = PointValue
    point = '0 0.5 0'
    variable = u
  []
  [initialfunc]
    type = FunctionValuePostprocessor
    function = conductivity
  []
[]
````

Output I am getting is:
Postprocessor Values:
sub20: +----------------+----------------+----------------+----------------+
sub20: | time                  | initialfunc        | left_bc              | left_bc5            |
sub20: +----------------+----------------+----------------+----------------+
sub20: |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
sub20: |   1.000000e-02 |   5.000000e+00 |   8.970800e+00 |   8.970800e+00 |
sub20: |   2.000000e-02 |   5.000000e+00 |   8.970800e+00 |   8.970800e+00 |
sub20: |   3.000000e-02 |   5.000000e+00 |   8.970800e+00 |   8.970800e+00 |
sub20: |   4.000000e-02 |   5.000000e+00 |   8.970800e+00 |   8.970800e+00 |
sub20: |   5.000000e-02 |   5.000000e+00 |   8.970800e+00 |   8.970800e+00 |
sub20: +----------------+----------------+----------------+----------------+

I am looking for left_bc to vary along time randomly and also left_bc5 should be randomly distributed along time and should not be same as left_bc

**Reply by jiangwen84** [reply]
```
[StochasticTools]
  auto_create_executioner = false
[]

[Distributions]
  [node1_uniform]
    type = Uniform
    lower_bound = 5
    upper_bound = 10
  []
  [node2_uniform]
    type = Uniform
    lower_bound = 5
    upper_bound = 10
  []
  [node3_uniform]
    type = Uniform
    lower_bound = 5
    upper_bound = 10
  []
[]

[Samplers]
  [sample]
    type = MonteCarlo
    num_rows = 5
    distributions = 'node1_uniform node2_uniform node3_uniform'
    execute_on = 'PRE_MULTIAPP_SETUP'

  []
[]

[MultiApps]
  [sub2]
    type = SamplerTransientMultiApp
    sampler = sample
    input_files = 'sub2.i'
    execute_on = 'INITIAL TIMESTEP_END'
  []
[]

[Controls]
  [cmdline]
    type = MultiAppSamplerControl
    multi_app = sub2
    sampler = sample
    param_names = 'node1_value node2_value node3_value'
  []
[]

[Executioner]
  type = Transient
  num_steps = 5
  dt = 0.01
[]
```


```
node1_value = 0.5
node2_value = 0.5
node3_value = 0.5

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 10
  ny = 10
[]

[Variables]
  [u]
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = u
  []
  [time]
    type = TimeDerivative
    variable = u
  []
[]

[BCs]
  [left]
    type = FunctionDirichletBC
    variable = u
    boundary = left
    function = conductivity
  []
  [right]
    type = DirichletBC
    variable = u
    boundary = right
    value = 1
  []
  [top]
    type = NeumannBC
    variable = u
    boundary = top
    value = 0
  []
  [bottom]
    type = NeumannBC
    variable = u
    boundary = bottom
    value = 0
  []
[]

[Functions]
  [conductivity]
    type = PiecewiseLinear
    axis = y
    x = '0 0.5 1'
    y = '${node1_value} ${node2_value} ${node3_value}'
  []
[]

[Executioner]
  type = Transient
  num_steps = 5
  dt = 0.01
  solve_type = PJFNK
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Controls]
  [stochastic]
    type = SamplerReceiver
  []
[]

[Postprocessors]
  [left_bc]
    type = PointValue
    point = '0 0 0'
    variable = u
  []
  [left_bc5]
    type = PointValue
    point = '0 0.5 0'
    variable = u
  []
  [initialfunc]
    type = FunctionValuePostprocessor
    function = conductivity
  []
[]

[Outputs]
  csv = true
[]
```

**Reply by vermaprk** [reply]
Thanks. This works well with y nodes.
But I also want the randomness with time. In the output, values at all time steps are same. 
Postprocessor Values:
sub24: +----------------+----------------+----------------+----------------+
sub24: | time           | initialfunc    | left_bc        | left_bc5       |
sub24: +----------------+----------------+----------------+----------------+
sub24: |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
sub24: |   1.000000e-02 |   5.637317e+00 |   5.637317e+00 |   9.273206e+00 |
sub24: |   2.000000e-02 |   5.637317e+00 |   5.637317e+00 |   9.273206e+00 |
sub24: |   3.000000e-02 |   5.637317e+00 |   5.637317e+00 |   9.273206e+00 |
sub24: |   4.000000e-02 |   5.637317e+00 |   5.637317e+00 |   9.273206e+00 |
sub24: |   5.000000e-02 |   5.637317e+00 |   5.637317e+00 |   9.273206e+00 |
sub24: +----------------+----------------+----------------+----------------+

In the Samplers I tried to execute_on at timestep_begin but it doesn't works other than PRE_MULTIAPP_SETUP
```
[Samplers]
  [sample]
    type = MonteCarlo
    num_rows = 1
    distributions = 'node1_uniform node2_uniform node3_uniform'
    execute_on = 'PRE_MULTIAPP_SETUP'
  []
```
Thanks

**Reply by jiangwen84** [reply]
The limitation is from ```MultiAppSamplerControl``` which only works for ```PRE_MULTIAPP_SETUP``` for now.  You can try different MultiApp control methods.

**Reply by vermaprk** [reply]
ok. Thanks a lot
